crates/fixedpointmath/src/fixed_point.rs

```diff
@@ -0,0 +1,399 @@
+use std::fmt::{self, Debug};
+
+use ethers::types::{I256, U256};
+use eyre::{bail, eyre, Result};
+
+use crate::{sign::FixedPointSign, utils::u256_from_str, value::FixedPointValue};
+
+/// A generic fixed point wrapper around the `U256` type from ethers-rs.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct FixedPoint<T: FixedPointValue> {
+    raw: T,
+    decimals: u8,
+}
+
+impl<T: FixedPointValue> FixedPoint<T> {
+    pub const MIN: Self = Self {
+        raw: T::MIN,
+        decimals: T::MAX_DECIMALS,
+    };
+
+    pub const MAX: Self = Self {
+        raw: T::MAX,
+        decimals: T::MAX_DECIMALS,
+    };
+
+    // Constructors //
+
+    pub fn new<V: Into<T>>(value: V) -> Self {
+        Self {
+            raw: value.into(),
+            // TODO: Add support for variable decimals.
+            decimals: T::MAX_DECIMALS,
+        }
+    }
+
+    pub fn try_from<V: TryInto<T> + Debug>(value: V) -> Result<Self> {
+        // Convert the value to a Debug string before moving it incase the
+        // conversion fails.
+        let value_debug = format!("{:?}", value);
+        let value = value.try_into().map_err(|_| {
+            eyre!(
+                r#"Failed to convert value to underlying FixedPointValue type:
+    value: {value_debug}
+    Underlying range: {min:?} to {max:?}
+"#,
+                min = Self::MIN,
+                max = Self::MAX,
+            )
+        })?;
+
+        Ok(Self::new(value))
+    }
+
+    pub fn from_sign_and_abs(sign: FixedPointSign, abs: U256) -> Result<Self> {
+        Ok(match sign {
+            FixedPointSign::Positive => Self::new(T::from_u256(abs)?),
+            FixedPointSign::Negative => {
+                if abs == T::MIN.unsigned_abs() {
+                    // Avoid overflow
+                    Self::MIN
```

Comment:
Trying to call `T::from_u256(T::MIN.unsigned_abs())` would overflow since the
absolute `MIN` value of a two's compliment integer is 1 more than its `MAX`, so instead
of creating the positive value first, then flipping the sign, I'm just
shortcutting to the actual value.

I'll add a comment for this: 👍